### PR TITLE
Fix bugged completion of ~/some in vi mode

### PR DESCRIPTION
**Pull request by JohnoKing**
This pull request fixes the bug reported in att/ast#682. The following sequence fails in vi mode because ksh looks in the wrong part of the `virtual` buffer:

```sh
$ touch ~/testfile
$ ls ~/test<tab>
```

This is fixed by changing `virtual[i]` to `virtual[last_virt]` in the bugged section of code. The other change in this pull request is to make sure listing files in a directory with something like `ls /etc/<tab>` calls the code for Ctrl+L to preserve `ls /etc/` rather than try (and fail) to complete the directory name, producing `ls /etc\n/`. This bugfix was backported from ksh93v- 2013-10-10-alpha. The ksh93v- changelog list a different example for this bug:

>    12-08-31  A bug in vi mode in which ~/tst/make/users/hv<TAB> did not expand correctly has been fixed

~There was also a minor typo in the NEWS file that I fixed as part of this pull request.~

### CODE REVIEW

File: NEWS [hunk 1]
```diff
@@ -45,8 +51,8 @@ Any uppercase BUG_* names are modernish shell bug IDs.
 
 - A two decade old bug that caused 'whence -a' to base the path of
   tracked aliases on the user's current working directory has been
-  fixed. Now the real path to the tracked aliases is shown when '-a'
-  is passed to the whence command.
+  fixed. Now the real path to tracked aliases is shown when '-a' is
+  passed to the whence command.
```
Comment (McDutchie) [inline NEWS]: You might have noticed that, in `README.md`, I linked the `NEWS` file to GitHub's version of `git blame`, so that people can easily associate news items with commits revealing the full nitty gritty. So I would prefer to have no trivial edits of items that are good enough, as that would confuse this system. If there are serious errors in `NEWS`, they should be fixed in their own commits.

**Review by McDutchie — COMMENTED** [review]
Funny. I can't reproduce this bug on my Mac at all, either on stock 93u+ or on 93u+m. But I can reproduce it on (at least) Linux, FreeBSD, and HP-UX. And on those systems this bug is *not* limited to vi mode, ~~it shows up in emacs mode as well.~~ *it also shows up when both vi mode and emacs mode are disabled.*

And there is another symptom on those systems: *if both vi mode and emacs mode are disabled*, after triggering this bug once, subsequent tab completions produce an extra character at the end of the command line (being the character that was swallowed by the bug) that is impossible to delete. Wow.

Anyway... I'll test this on various systems for a bit before merging this, if you don't mind. We should also figure out a regression test for this bug; it would use pty and go in pty.sh.

Meanwhile I've got one comment on editing the `NEWS` file, see below.

File: NEWS (hunk 1 above)
Comment (JohnoKing) [reply NEWS]: That makes sense. The typo fix is no longer a part of the pull request.

**Comment by JohnoKing** [comment]
I can't say anything about HP-UX, but I can't reproduce this bug in emacs mode on Linux or FreeBSD.

**Comment by posguy99** [comment]
I’m sure you saw this one, but it would seem that https://github.com/att/ast/issues/1412 is also relevant, and probably why it’s not necessarily reproducible on macOS.

Having said that, I can’t tell you how to reproduce it, but I *have* seen @McDutchie’s example of tab completions in emacs mode producing an extra character you can’t delete, but I can’t tell you how to trip it.

macOS 93u+ 2012-08-01

--
Marc Wilson
posguy99@gmail.com

**Comment by JohnoKing** [comment]
I have managed to write a `pty` regression test for this bug, but the test is slightly different from the original reproducer. If the home directory is under `/tmp`, the string length must be five characters for the bug to be reproducible, with the test case being:

```sh
$ mkdir /tmp/fakehome
$ HOME=/tmp/fakehome
$ touch ~/testfile
$ ls ~/tes<tab>
```

Running the same sequence above without an editing mode enabled causes ksh to print four newlines.

> And there is another symptom on those systems: after triggering this bug once, subsequent tab completions produce an extra character at the end of the command line (being the character that was swallowed by the bug) that is impossible to delete.

I still can't get this to reproduce in emacs mode, but I can replicate it when all editing modes are disabled: 

```sh
$ set +o vi; set +o emacs; set +o gmacs
$ touch ~/testfile
$ ls ~/testf<tab>
```

File: src/cmd/ksh93/tests/pty.sh [hunk 2]
```diff
@@ -453,4 +453,21 @@ w fg
 u yes-yes
 !
 
+# ======
+# Test file name completion in vi mode
```
Comment (McDutchie) [inline src/cmd/ksh93/tests/pty.sh:458]: Please add a dummy `# err_exit #` comment like on the other tests, so that shtests counts it.

**Comment by McDutchie** [comment]
Indeed, I must have made a mistake in assuming emacs mode was active while in fact both editing modes were disabled. I've edited my comment above.

The pull request fixes vi mode, and it fixes the same bug in "neither" mode, but I've found that it leaves that other bug in "neither" mode: the undeleteable extra character.

@JohnoKing, do you want to take a stab at fixing that as part of this pull request, or leave it for another issue/PR? Either way is fine with me.

~~I'd also love to know why the Mac seems to be immune to both of these bugs…~~ Never mind. It isn't. The triggers are slightly different, though.

**Comment by JohnoKing** [comment]
I'd rather fix the no editing mode problem in a different PR, although FWIW I have done some debugging and have found that bug doesn't occur in ksh93q 1993-12-28.

**Comment by McDutchie** [comment]
Note that what @krader1961 reports in https://github.com/att/ast/issues/1412#issuecomment-539296120 is normal and expected behaviour, because `a` is the only file in the just-created `test` directory.

**Comment by McDutchie** [comment]
Did a force-push to master to add a NEWS file entry, sorry if that inconvenienced you.

**Comment by McDutchie** [comment]
Trying out the "neither" mode on various systems, I'm now finding it unpredictable whether there is any tab completion at all. Sometimes I get broken tab completion, sometimes none. I suspect the use of an uninitialised variable somewhere.

Reading the manual page, tab completion is only documented for "each of the editing modes", so it seems best to simply make sure tab completion is properly deactivated when neither emacs nor vi mode is active.
